text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,451 @@
+1. **complexity** - We tried to be clever and support appending argument appending to shell processes. This "works" but can result in surprising behavior such as the need to escape literal `"` characters to prevent Bash from removing them during evaluation. Users must know whether the process is a direct process or a shell process in order to provide arguments correctly, making process types a leaky abstraction.
+2. **use cases** - Arguments can serve two different purposes. Some are required and should always be included in the command. Some are merely defaults that may be overridden by the user. The current spec does not support the latter case.
+
+## Shell-Specific Logic
+
+**interoperability** - Currently, unless a buildpack author specifically indicates that the given process type is a `direct` process, a shell process with a dependency on Bash (linux) or Cmd (windows) is created. This is not ideal, as many in the industry are moving towards minimal images that do not include a shell, in order to reduce surface attack area. Buildpack authors may inadvertantly create process types that depend on a shell even when no such dependency is necessary.
+
+**complexity** - Having a direct dependency between the launcher and specific shells is inelegant. It makes the spec more complex and end users and buildpack author are forced to understand that complexity in order to understand the behavior of the resulting usage or debug issues. For example end users and buildpacks authors must understand:
+- The difference between a direct and shell process
+- The nuances of argument handling in direct vs. Bash vs. Cmd cases
+- That buildpack-provided profile scripts will not apply to direct processes
+- That a user-provided `.profile` script will not apply to a direct processes
+
+ Removing this special behavior and requiring buildpacks to explicitly include any required shell in the command itself creates a simpler, more comprehensible model for buildpack authors and end users alike.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Process Schema
+The existing schema for the processes table in `launch.toml` is
+```
+# Old Schema
+
+[[processes]]
+type = "required"
+command = "required"
+args = ["optional"]
+direct = false
+default = false
+```
+
+The new proposed schema is:
+```
+# New Schema
+
+[[processes]]
+type = "required"
+command = ["required"]
+args = ["optional"]
+default = false
+```
+The following changes have been made:
+- **`direct` has been removed** - all processes are executed directly. If `bash` or `cmd.exe` is required it should be included in `command`. No surprises.
```

Comment:
I think there could be some nuance here that we need to document when switching from a shell process to direct regarding signals. Many platforms support the concept of a graceful shutdown, which is a SIGTERM followed by a pause, followed by SIGKILL on your process.

When you run as a shell process, pid 1 is your shell which generally traps and handles signals as you'd expect. When you switch to running your process as direct, it becomes pid 1 and it may or may not handle all the signals correctly (you can generally tell if it's not handling them correctly by running a container and trying to interrupt it with ctrl+c, if that works it's handling signals correctly). 

Anyway, I think we need to call out this consideration so that when buildpack authors are migrating their buildpacks they think about this nuance and ensure it's handled correctly.

Reply:
After some reading I suspect the default signal handling behavior is actually better for a direct process than a shell process. Bash does not forward the SIGTERM to the child process by default and therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. Therefore a buildpack author that is not thinking about this explicitly might prevent the appropriate handling of a sigterm if they default to creating a shell process.

https://veithen.io/2014/11/16/sigterm-propagation.html

cc @hone, @dmikusa-pivotal

Reply:
> therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. 

Surprisingly, it's even worse than this. A simple bash invocation (e.g., `bash -c "tail -f /dev/null"`) is treated differently by bash than a complex one (e.g., `bash -c "true && tail -f /dev/null"`). In the former case, bash `exec`s itself away, resulting in a single process. So from the user's perspective, whether this works or not is quite arbitrary.

Reply:
Hmm, why isn't anything ever easy?  Lol.

The behavior I'd observed is with apps that don't handle signals at all. If you run them as PID1, you can't terminate the container. It's SIGKILL or the app keeps running. However, if you wrap that in `bash -c` it'll terminate as expected. I guess I didn't dig deep enough to check how Bash was in turn stopping that app, based on what's linked here it is just killing it as well and not propagating signals.

Regardless, there seems to be enough complexity here that I still think this warrants a mention in this RFC document. It's nuanced behavior that could easily be missed by a buildpack author, if we call attention to it, that gives authors a chance to investigate what it means to them. 

Even if it's as simple as "When running a process as direct, your process will always run as PID1. Buildpack authors, please be aware this may change the behavior of your application at shutdown and you should review accordingly.", I think that give's authors a heads up.

Reply:
One more "fun" point on this subject. If you use `sh -c` instead of `bash -c` you get more different behavior because at least on Ubuntu, `/bin/sh` is not `/bin/bash`, it's `/bin/dash`. I was running a process this morning with `sh -c` and it completely ignored SIGTERM.

Reply:
@sclevine Are you aware of any documentation regarding when bash does or doesn't exec the child process. When I tried out you example commands I observed Bash replacing itself in memory, even for the more complex command:
```
docker run --rm --name test --entrypoint bash ubuntu:bionic -c 'true && tail -f /dev/null'
```
```
docker exec test ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
root           1       0  0 14:41 ?        00:00:00 tail -f /dev/null
root           7       0  0 14:41 ?        00:00:00 ps -ef
```

Reply:
Okay, if I am understanding correctly, I think there are relevant two cases here:
1. the command is a complicated Bash invocation
2. the command is a simple Bash invocation

My understanding of the behavior (based on @sclevine 's description and my experimental verification):

**case 1**: A complicated command cannot be executed directly (`true && tail -f /dev/null` is not a valid direct process) and therefore `bash -c` must be explicitly included in the process. Therefore, wrt signal handling, **nothing changes**.
**case 2**: Bash is already `exec`ing itself away leaving the process as PID1. If this is converted to a direct process the same process will be PID1. Therefore, wrt signal handling, **nothing changes**.

**Summary**: With regards to signal handling nothing changes. App developers and buildpacks authors do not need to worry about changes to signal handling.

Reply:
@dmikusa-pivotal & @hone do you feel that this comment can be resolved given the explanation above describing how this does not affect end users or buildpack authors?

Reply:
I think this discussion has been good around signals. It seems unlikely that it's going to cause issues, so I think we're good there.

Maybe one final suggestion with this sentence...

>If `bash` or `cmd.exe` is required it should be included in `command`.

I think "required" might be nebulous and assumes you know what is required. It's easy to overlook restrictions of running as direct/PID1. Perhaps we could have something like:

>If currently using `direct = false`, the recommended migration path for buildpack authors is to include `bash` or `cmd.exe`  in `command`. This will result in the same behavior as `direct = false`. Before switching to `direct = true`, it is recommended you confirm environment variable interpolation is not required and that your process can function as PID1.

I guess it doesn't necessarily have to be in this line, but I think we should offer some migration advice and list the concerns an author should review before switching to `direct = true`.

Reply:
@dmikusa-pivotal please see https://github.com/buildpacks/rfcs/pull/168/commits/14066e50a3ee999fe0ff5249362e43c52169cac5

Reply:
Looks great, thanks!